src/cmd/load-debug-dump-into-postgres/main.go

```diff
@@ -0,0 +1,290 @@
+			// Perhaps this is a pgbouncer log; we will try to parse it.
+			t, f, ok := parsePgBouncerLine(line)
+			if ok {
+				ts.Valid = true
+				ts.Time = t
+				parsed = f
+				parseError.Valid = false
+			}
+		}
+
+		tag, err := db.Exec(ctx, "insert into logs(hash, dumpname, filename, time, line, parsed, parseerror) values ($1, $2, $3, $4, $5, $6, $7) on conflict do nothing", hashBytes, *source, name, ts, line, parsed, parseError)
+		if err != nil {
+			if err := context.Cause(ctx); err != nil {
+				// This is where we exit when C-c is pressed.
+				nerr.Add(1)
+				return err //nolint:wrapcheck
+			}
+			log.Info(ctx, "inserting line failed", zap.Error(err))
+		}
+		switch tag.RowsAffected() {
+		case 0:
+			ndups.Add(1)
+		default:
+			nrows.Add(tag.RowsAffected())
+		}
+	}
+	if err := s.Err(); err != nil {
+		return errors.Wrap(err, "scan")
+	}
+	return nil
+}
+
+func printStats(ctx context.Context) {
+	stats := func(m string) {
+		log.Info(ctx, m, zap.Int64("duplicates", ndups.Load()), zap.Int64("rows", nrows.Load()), zap.Int64("json", njson.Load()), zap.Int64("lines", nlines.Load()), zap.Int64("files", nfiles.Load()), zap.Int64("binary", nbinary.Load()), zap.Int64("bytes_read", nbytesin.Load()), zap.Int64("bytes_processed", nbytesout.Load()), zap.Int64("bytes_in_flight", nbytesin.Load()-nbytesout.Load()), zap.Int64("running", ngoro.Load()), zap.Int64("awaiting_db", nconnacq.Load()))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			stats("final stats")
+			return
+		case <-time.After(time.Second):
+			stats("stats")
+		}
+
+	}
+}
+
+func main() {
+	flag.Parse()
+	done := log.InitBatchLogger("")
+	defer func() {
+		var err error
+		if nerr.Load() > 0 {
+			err = errors.New("dump errored")
+		}
+		done(err)
+	}()
+
+	if *v {
+		log.SetLevel(log.DebugLevel)
+	}
+
+	ctx, c := pctx.Interactive()
+
+	pool, err := pgxpool.Connect(ctx, *dsn)
+	if err != nil {
+		log.Exit(ctx, "problem connecting to database", zap.Error(err))
+	}
+
+	fr, err := os.Open(*source)
+	if err != nil {
+		log.Exit(ctx, "problem opening source", zap.Error(err))
+	}
+	gr, err := gzip.NewReader(fr)
+	if err != nil {
+		log.Exit(ctx, "problem creating gzip reader", zap.Error(err))
+	}
+
+	go printStats(ctx)
+
+	wg := new(sync.WaitGroup)
+	tr := tar.NewReader(gr)
+	for {
+		h, err := tr.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			// but wait for everything else to finish
+			log.Error(ctx, "problem reading dump", zap.Error(err))
+			nerr.Add(1)
+			break
+		}
+
+		var buf bytes.Buffer
+		n, err := io.Copy(&buf, tr)
+		if err != nil {
+			log.Debug(ctx, "read error", zap.String("filename", h.Name), zap.Error(err))
+			continue
+		}
+		nbytesin.Add(n)
+
+		nconnacq.Add(1)
+		c, err := pool.Acquire(ctx)
+		nconnacq.Add(-1)
+		if err != nil {
+			log.Error(ctx, "cannot acquire db conn", zap.Error(err))
+			nerr.Add(1)
+			break
+		}
+
+		wg.Add(1)
+		go func() {
+			addFile(ctx, &buf, h.Name, c) //nolint:errcheck
```

Comment:
You _could_ collect errors … not a huge deal in a quick hack.

Reply:
They're in the log; this causes the log to be retained and its filename to be printed.